Sources/GeneralKit/Services/GithubService/GithubServiceImpl.swift

```diff
@@ -0,0 +1,137 @@
+
+    private let dependencies: Dependencies
+
+    public init(dependencies: Dependencies) {
+        self.dependencies = dependencies
+    }
+
+    public func getGitRepoPath(repo: String) throws -> String {
+        let components = repo.split(separator: " ")
+        guard let name = components.first else {
+            throw Error.githubName(repo)
+        }
+        var branch = Constants.defaultGithubBranch
+        if components.count > 1 {
+            branch = String(components[1])
+        }
+
+        return Constants.githubArchivePath(String(name), branch)
+    }
+
+    public func downloadFiles(at repo: String, filesHandler: ([FileInfo]) throws -> Void) throws {
+        let gitRepoPath = try getGitRepoPath(repo: repo)
+        let archiveURL = try downloadArchive(at: gitRepoPath)
+        let folderURL = try dependencies.fileHelper.unzipArchive(at: archiveURL)
+        let firstDirectory = try dependencies.fileHelper.contentsOfDirectory(at: folderURL).first(where: \.isDirectory)
+        guard let directory = firstDirectory else {
+            throw Error.write(folderURL)
+        }
+        let files = try dependencies.fileHelper.contentsOfDirectory(at: directory.url)
+        do {
+            try filesHandler(files)
+        }
+        catch {
+            try dependencies.fileHelper.removeFile(at: folderURL)
+            throw error
+        }
+        try dependencies.fileHelper.removeFile(at: folderURL)
+    }
+
+    @discardableResult
+    public func downloadFiles(at repo: String,
+                              to destination: String) throws -> [FileInfo] {
+        var movedFiles: [FileInfo] = []
+        let destination = URL(fileURLWithPath: destination)
+        try downloadFiles(at: repo) { files in
+            movedFiles.append(contentsOf: try move(files, to: destination))
+        }
+        return movedFiles
+    }
+
+    // MARK: - Private
+
+    private func downloadArchive(at path: String) throws -> URL {
+        guard let url = URL(string: path) else {
+            throw Error.url(path)
+        }
+        let data = try self.data(at: url)
+        let tmpFolderURL = try URL.tempFolderWithRandomName()
+        let fileURL = tmpFolderURL + url.lastPathComponent
+        do {
+            try data.write(to: fileURL)
+        }
+        catch {
+            throw Error.write(tmpFolderURL)
+        }
+        return fileURL
+    }
+
+    private func data(at url: URL) throws -> Data {
+        do {
+            return try Data(contentsOf: url)
+        }
+        catch {
+            throw Error.download(url)
+        }
+    }
+
+    private func fetchFiles(in url: URL, matchHandler: (FileInfo) -> Bool) throws -> [FileInfo] {
+        var files = [FileInfo]()
+        try dependencies.fileHelper.contentsOfDirectory(at: url).forEach { file in
+            if matchHandler(file) {
+                files.append(file)
+            }
+        }
+        return files
```

Comment:
map is faster then for-loop ([source](https://www.skoumal.com/en/performance-of-built-in-higher-order-functions-map-filter-reduce-and-flatmap-vs-for-in-loop-in-swift/)). So I think compactMap will be faster here too 🤓

Reply:
don't see how it can be pure rewritten to `compactMap`. `reduce` here is more applicable, but according your source, for-in is faster than `reduce`